### PR TITLE
Add FxA Pairing flow methods

### LINKED ISSUE #237: Add support for launching a pairing-based OAuth flow
We need to figure out the precise API to use here, but the idea would be to accept the data scanned out of a QR code from Desktop, and use it to launch into the web-content-hosted Supplicant pairing flow.

We currently have a method like `begin_oauth_flow(scopes, wants_keys)` that returns a URL for the app to open.  Would it suffice to just have a `begin_pairing_flow(pairind_data, scopes, wants_keys)` method that does a similar thing?

@vladikoff what approach did you take with your demo app?

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes #237 
Fixes https://github.com/mozilla/fxa/issues/312

### CODE REVIEW

**Comment by vladikoff** [comment]
@eoger r?

File: fxa-client/src/lib.rs [hunk 1]
```diff
@@ -294,6 +294,44 @@ impl FirefoxAccount {
         Ok(Some(self.handle_oauth_token_response(resp, None)?))
     }
 
+    pub fn begin_pairing_flow(&mut self, pairing_url: &str, scopes: &[&str]) -> Result<String> {
+        let mut url = self.state.config.content_url_path("/pair/supp")?;
+        let pairing_url = Url::parse(pairing_url)?;
```
Comment (vladikoff) [inline fxa-client/src/lib.rs]: @eoger the approach here is to hide the details of the whole `fragment` thing into the rust lib, so we don't have to do it in Kotlin, Swift, etc..

File: fxa-client/src/lib.rs [hunk 2]
```diff
@@ -294,6 +294,44 @@ impl FirefoxAccount {
...
+            .append_pair("code_challenge", &code_challenge)
+            .append_pair("access_type", "offline");
+
+        url.set_fragment(pairing_url.fragment());
+
+        let flow = ScopedKeysFlow::with_random_key(&*RNG)?;
+        let jwk_json = flow.generate_keys_jwk()?;
+        let scoped_keys_flow = Some(flow);
+        let keys_jwk = base64::encode_config(&jwk_json, base64::URL_SAFE_NO_PAD);
+        url.query_pairs_mut().append_pair("keys_jwk", &keys_jwk);
```
Comment (rfk) [inline fxa-client/src/lib.rs]: All of this setup logic already exists in `begin_oauth_flow`, is this an opportunity to refactor it into a shared helper of some sort?

File: fxa-client/src/lib.rs (hunk 1 above)
Comment (rfk) [reply fxa-client/src/lib.rs]: Please check that the `pairing_url` has the same origin as the content-serve from `self.state.config`.

File: fxa-client/src/lib.rs [hunk 3]
```diff
@@ -294,6 +294,44 @@ impl FirefoxAccount {
...
+        url.query_pairs_mut()
+            .append_pair("client_id", &self.state.client_id)
+            .append_pair("redirect_uri", &self.state.redirect_uri)
+            .append_pair("scope", &scopes.join(" "))
+            .append_pair("state", &state)
+            .append_pair("code_challenge_method", "S256")
+            .append_pair("code_challenge", &code_challenge)
+            .append_pair("access_type", "offline");
+
+        url.set_fragment(pairing_url.fragment());
```
Comment (eoger) [inline fxa-client/src/lib.rs]: Oh! hon hon hon 🥖 
PS: maybe put the `parse` next to the fragment so we can see what's going on.

**Comment by vladikoff** [comment]
Updated!

File: fxa-client/src/lib.rs [hunk 4]
```diff
@@ -294,18 +294,37 @@ impl FirefoxAccount {
         Ok(Some(self.handle_oauth_token_response(resp, None)?))
     }
 
+    pub fn begin_pairing_flow(&mut self, pairing_url: &str, scopes: &[&str]) -> Result<String> {
+        let mut url = self.state.config.content_url_path("/pair/supp")?;
+        let pairing_url = Url::parse(pairing_url)?;
+
+        if url.host_str() != pairing_url.host_str() {
+            return Err(ErrorKind::OriginMismatch.into());
+        };
```
Comment (eoger) [inline fxa-client/src/lib.rs]: ;;;;;;;;;;;

**Pull request merged**